research-prototype/Programs/Source/ipae2e.mpc

```diff
@@ -0,0 +1,154 @@
+import os
+
+from Compiler.library import print_ln
+from Compiler.types import sint, Matrix, Array
+
+
+numrows_power = int(os.getenv('IPA_NUMROWS_POWER', 10))
+numrows = 2 ** numrows_power
+
+verbose = os.getenv('IPA_VERBOSE', None)
+if verbose is not None:
+    if verbose == 'True':
+        verbose = True
+    elif verbose == 'False':
+        verbose = False
+    else:
+        raise Exception(f"Unknown verbose value: {verbose}")
+else:
+    verbose = False
+
+if verbose and numrows_power > 5:
+    raise Exception(
+        f"You cannot compiling with VERBOSE on for {numrows} rows. "
+        f"This will result in quite a bit of output in your terminal."
+    )
+
+print_ln('********************  IPA E2E %s **********************', numrows)
+
+_list = Matrix(numrows, 4, sint)
+
+local_params = {
+    '_list': _list,
+    'sint': sint,
+    'print_ln': print_ln,
+}
+
+for i in range(numrows):
+    for j in range(4):
+        cmd = f"_list[{i}][{j}] = sint.get_input_from(0)"
+        exec(cmd, {}, local_params)
+
+# Radix sort
+_list.sort((0,), n_bits=32)
+
+if verbose:
+    for i in range(numrows):
+        cmd = (f"print_ln('%s, %s, %s, %s', "
+               f"_list[{i}][0].reveal(), _list[{i}][1].reveal(),"
+               f"_list[{i}][2].reveal(), _list[{i}][3].reveal(),)")
+        exec(cmd, {}, local_params)
+
+helperbit = Array(numrows, sint)
+local_params['helperbit'] = helperbit
+
+for i in range(numrows-1):
+    cmd = f"helperbit[{i}] = (_list[{i}][0] == _list[{i}+1][0]).if_else(1, 0)"
+    exec(cmd, {}, local_params)
+
+
+helperbit_and_istrigger = Array(numrows, sint)
+local_params['helperbit_and_istrigger'] = helperbit_and_istrigger
+helperbit_xor_istrigger = Array(numrows, sint)
+local_params['helperbit_xor_istrigger'] = helperbit_xor_istrigger
+
+for i in range(numrows-1):
+    cmd = f"helperbit_and_istrigger[{i}] = helperbit[{i}] * _list[{i}][1]"
+    exec(cmd, {}, local_params)
+    cmd = f"helperbit_xor_istrigger[{i}] = (helperbit[{i}] + _list[{i}][1]) % 2"
+    exec(cmd, {}, local_params)
+
+if verbose:
+    print_ln('%s', helperbit.reveal())
+    print_ln('%s', helperbit_and_istrigger.reveal())
+    print_ln('%s', helperbit_xor_istrigger.reveal())
+
+
+stopbits = [Array(numrows, sint) for _ in range(numrows_power)]
+local_params['stopbits'] = stopbits
+
+for stopbit in stopbits:
+    local_params['stopbit'] = stopbit
+    for i in range(numrows):
+        cmd = f"stopbit[{i}] = 1"
+        exec(cmd, {}, local_params)
+
+credits = [Array(numrows, sint) for _ in range(numrows_power)]
+local_params['credits'] = credits
+
+for i, (stopbit, credit) in enumerate(zip(stopbits, credits)):
```

Comment:
What does enumerate(zip(a, b)) do?

Reply:
`enumerate(['a', 'b', 'c']) -> [(0, 'a'), (1, 'b'), (2, 'c')]`. See the [python docs.](https://docs.python.org/3/library/functions.html#enumerate).

`zip([10, 20, 30], [100, 200, 300]) -> [(10, 100), (20, 200), (30, 300)]`. See the [python docs.](https://docs.python.org/3/library/functions.html#zip).